✨🤖 Update ConvE to ERModel

mberr opened the pull request:
This PR updates `ConvE` to new-style `ERModel`.

#### Benchmark

**TL;DR**: training is a bit slower (new `18.37`s vs. `17.41`s); evaluation roughtly the same (new `1.72`s vs. `1.70`s).

(possible) reasons
* [x] unnecessary selection of entity biases for head entities, cf ef74c55 -> minor impact

<details>

| Key             | Value                    |
|-----------------|--------------------------|
| OS              | posix                    |
| Platform        | Linux                    |
| Release         | 5.4.0-81-generic         |
| Time            | Sat Apr 16 18:12:51 2022 |
| Python          | 3.8.10                   |
| PyKEEN          | 1.8.1-dev                |
| PyKEEN Hash     | *                 |
| PyKEEN Branch   | *                   |
| PyTorch         | 1.10.0+cu113             |
| CUDA Available? | true                     |
| CUDA Version    | 11.3                     |
| cuDNN Version   | 8200                     |

```console
pykeen experiments run dettmers2018_conve_fb15k237.json
```
with the configuration being equal to https://github.com/pykeen/pykeen/blob/f47ee602573c87f6dace277ff656d2b12eeb5430/src/pykeen/experiments/conve/dettmers2018_conve_fb15k237.json, except that the number of training epochs has been reduced to 10.

| commit | train | eval |
| -- | -- | -- |
| f47ee602573c87f6dace277ff656d2b12eeb5430 (`master`) | 17.41s | 1.70s* |
| b6024ce1ac78d87b823d7bd310727e9347687cdc  | 18.43s | 1.73s* |
| ef74c55a8857868393d3bfa2ad59b65ae390b430 (`update-conve-to-ermodel`) | 18.37s | 1.72s* |

* `*` with `batch_size=4096`

</details>

mberr committed:
update ConvE to ERModel

mberr committed:
fix ConvE interaction module

trigger ci

mberr committed:
fix slow inner product

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -652,7 +652,7 @@ def _prepare_hrt_for_functional(
         t: TailRepresentation,
     ) -> MutableMapping[str, torch.FloatTensor]:  # noqa: D102
-        return dict(h=h, r=r, t=t[0], t_bias=t[1])
+        return dict(h=h[0], r=r, t=t[0], t_bias=t[1])
```
here, we select entity biases for the head, although they are never used. This should be adapted in the base model, but becomes apparent here, since ConvE is the only model with tail biases (afaik).

cthoyt replied:
Wait there’s at least one other model that has head and tail specific representations

mberr replied:
`CP` uses separate head / tail representations.

mberr committed:
enable CP / ConvE to use a subset of ent. repr.

mberr committed:
fix type annotation

mberr committed:
move index information into interaction

trigger ci

mberr committed:
trigger ci

mberr committed:
trigger ci

cthoyt commented on src/pykeen/nn/modules.py:
```diff
@@ -148,6 +150,43 @@ class Interaction(nn.Module, Generic[HeadRepresentation, RelationRepresentation,
+    def tail_indices(cls) -> Sequence[int]:
+        """Return the entity representation indices used for the tail representations."""
+        if cls._tail_indices is None:
+            tail_entity_shape = cls.entity_shape if cls.tail_entity_shape is None else cls.tail_entity_shape
```
Theres some asymmetry between the head and tail entity index functions

mberr replied:
cf. 553f4b8

cthoyt commented on src/pykeen/nn/modules.py:
```diff
@@ -148,6 +150,43 @@ class Interaction(nn.Module, Generic[HeadRepresentation, RelationRepresentation,
+        return cls._tail_indices
+
+    @classmethod
+    def full_entity_shapes(cls) -> Sequence[str]:
```
if this value is constant for the lifetime of the class, maybe it should be calculated in `__init_subclass__`? Though that has a bit of a problem with deep inheritance hierarchies. Maybe it can be cached, since this gets called quite a bit

cthoyt replied:
oh nvm this is only called once.

mberr committed:
Merge branch 'master' into update-conve-to-ermodel

mberr committed:
add comment
trigger ci

mberr committed:
Merge branch 'master' into update-conve-to-ermodel

mberr committed:
remove artifact

trigger ci

mberr committed:
fix conve interaction test tail bias shape

mberr committed:
convert classmethod

AutoSF takes a variable number of representations are input depending
on the coefficients it receives upon instantiation

mberr committed:
trigger ci

mberr committed:
trigger ci

mberr committed:
simplify code

trigger ci
